lib/db/observed.go

```diff
@@ -0,0 +1,253 @@
+	if err == nil {
+		err = db.Put(key, bs)
+	}
+	if err != nil {
+		l.Warnf("Failed to store pending device entry: %v", err)
+	}
+	return err
+}
+
+// PendingDevices lists unknown devices that tried to connect.  As a
+// side-effect, any invalid entries are dropped from the database
+// after a warning log message.  That's the only possible "repair"
+// measure and appropriate for the importance of pending entries.
+// They will come back soon if still relevant.
+func (db *Lowlevel) PendingDevices() (map[protocol.DeviceID]ObservedDevice, error) {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingDevice})
+	if err != nil {
+		return nil, err
+	}
+	res := make(map[protocol.DeviceID]ObservedDevice)
+	defer iter.Release()
+	for iter.Next() {
+		bs, err := db.Get(iter.Key())
+		if err != nil {
+			return nil, err
+		}
+		var od ObservedDevice
+		deviceID := db.keyer.DeviceFromPendingDeviceKey(iter.Key())
+		//FIXME: DeviceIDFromBytes() panics when given a wrong length input.
+		//       It should rather return an error which we'd check for here.
+		if len(deviceID) != protocol.DeviceIDLength {
+			goto deleteKey
+		}
+		if err := od.Unmarshal(bs); err != nil {
+			goto deleteKey
+		}
+		res[protocol.DeviceIDFromBytes(deviceID)] = od
+		continue
+	deleteKey:
+		l.Warnln("Invalid pending device entry, deleting from database")
+		if err := db.Delete(iter.Key()); err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+func (db *Lowlevel) RemovePendingDevice(device protocol.DeviceID) {
+	key := db.keyer.GeneratePendingDeviceKey(nil, device[:])
+	if err := db.Delete(key); err != nil {
+		l.Warnf("Failed to remove pending device entry: %v", err)
+	}
+}
+
+func (db *Lowlevel) AddOrUpdatePendingFolder(id, label string, device protocol.DeviceID) error {
+	key, err := db.keyer.GeneratePendingFolderKey(nil, []byte(id), device[:])
+	if err == nil {
+		of := ObservedFolder{
+			Time:  time.Now().Round(time.Second),
+			Label: label,
+		}
+		bs, err := of.Marshal()
+		if err == nil {
+			err = db.Put(key, bs)
+		}
+	}
+	if err != nil {
+		l.Warnf("Failed to store pending folder entry: %v", err)
+	}
+	return err
+}
+
+// PendingFolders lists folders that we don't yet share with the
+// offering devices.  As a side-effect, any invalid entries are
+// dropped from the database after a warning log message.  That's the
+// only possible "repair" measure and appropriate for the importance
+// of pending entries.  They will come back soon if still relevant.
+func (db *Lowlevel) PendingFolders(device protocol.DeviceID) (map[string]map[protocol.DeviceID]ObservedFolder, error) {
+	iter, err := db.NewPrefixIterator([]byte{KeyTypePendingFolder})
+	if err != nil {
+		return nil, err
+	}
+	res := make(map[string]map[protocol.DeviceID]ObservedFolder)
+	defer iter.Release()
+	for iter.Next() {
+		bs, err := db.Get(iter.Key())
+		if err != nil {
+			return nil, err
+		}
+		if keyDev, ok := db.keyer.DeviceFromPendingFolderKey(iter.Key()); ok {
```

Comment:
If you use `goto`, you can check for `!ok` and then `goto deleteKey` to get rid of one level of nesting.

Personally I am not a fan of goto, though admittedly only because I am not used to it (aka no good reason) :)  
I'd just replace all the gotos with
```
l.Warnln("Invalid pending folder entry, deleting from database")
return nil, db.Delete(iter.Key())
```
or potentially define a func variable doing that and calling that instead to get rid of the duplication. As said, just a matter of preference.

Reply:
I think `goto` within such a limited scope is fine, as it makes the code more readable, almost like prose text.  Saying "if this happens, then delete key", without having to name a function deleteKey() which would only be valid within this function's scope.  Repeating the error message seems awkward to me.

Definitely better than nesting condition blocks though, I agree.  So thanks for this additional hint, adopted.

Reply:
Sorry that was too quick, there was a good reason for the extra block:

    lib/db/observed.go:112:9: goto deleteKey jumps over declaration of folderID